Dispatcher system

bill-fitch14 opened the pull request.

bobjacobsen commented:
The Windows CI tests had a failure, so rerunning to see if that's just a spurious event:
```
[ERROR] Failures: 
[ERROR]   SCWarrantTest.testWarrant:114 "warrant not done" did not occur in time
[ERROR] Tests run: 27161, Failures: 1, Errors: 0, Skipped: 929
```

bobjacobsen commented:
The static checks of format and correctness found one concern:
```
1. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/beantable/TransitTableAction.java (at line 151)
217
	Transit z = getBySystemName((String) sysNameList.get(row));
218
	                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
219
Unnecessary cast from String to String
```

@bill-fitch14 Could you remove that cast and push the change again? It'll get automatically picked up here and rerun the checks.  Thank you!

bobjacobsen commented:
Thanks!  That update worked.  Unfortunately, the CI runs in phases, so it doesn't show all the issues all at once.  The next phase pulled up another thing:
```
[INFO] There are 12 errors reported by Checkstyle 8.29 with /home/runner/work/JMRI/JMRI/checkstyle.xml ruleset.
[ERROR] jython/DispatcherSystem/RunDispatch.py:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/CreateGraph.py:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/sensorsAndIcons/large_icon.xml:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/sensorsAndIcons/small_icon.xml:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/sensorsAndIcons/label.xml:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/sensorsAndIcons/conditionals.xml:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/sensorsAndIcons/memory.xml:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/sensorsAndIcons/sensor.xml:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/sensorsAndIcons/logix.xml:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/sensorsAndIcons/wide_icon.xml:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem/sensorsAndIcons/memory_icon.xml:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
[ERROR] jython/DispatcherSystem.py:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
```

These are files that end right after the last character, instead of having a newline at the end of the last line (so that they end with a newline character).  

These kinds of things are pretty routine with new projects, not a problem.

The fix is to open those files in your editor, go the end, put the cursor at the very bottom, and insert a newline.  Then commit and push as before.  

Thanks!

bobjacobsen commented:
One more thing:  If you want to put something about this in the release note, so that people know about it, you can edit that into the help/en/releasenotes/current-draft-note.shtml file.  That'll then appear in the draft and final release note, etc.

bill-fitch14 commented:
Looks as though it's failing for some reason. Anything I can correct?

bobjacobsen commented:
When you close and reopen, we lose the logs.  Could you wait a bit after this run and I'll take a look?

bobjacobsen commented:
The Windows CI check is still being pedantic about a newline:

```
[INFO] --- maven-checkstyle-plugin:3.1.0:check (default) @ jmri ---
464
[INFO] There is 1 error reported by Checkstyle 8.29 with /home/runner/work/JMRI/JMRI/checkstyle.xml ruleset.
465
[ERROR] jython/DispatcherSystem.py:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
```

bill-fitch14 commented:
What is the best way to do platform independent help. Is there an example in JMRI?
When does 'the file open dialog when the panel file has already been loaded' occur please.

bobjacobsen commented:
Most of JMRI's help is under the help/en/html directory (for general topics) or help/en/packages (for frame-specific help, organized by the classname of the frame).    You could make a new help/en/html/tools/dispatchersystem directory and putting things there.

Alternately, you could make a jython/DispatcherSystem/help directory and put things there.  We can still link those from the main index, etc.

dsand47 commented:
> What is the best way to do platform independent help. Is there an example in JMRI?
> When does 'the file open dialog when the panel file has already been loaded' occur please.

I would add a help page to the Dispatcher section at **help/en/package/jmri/jmrit/dispatcher/** with a link from the **Further Reference** section at https://www.jmri.org/help/en/package/jmri/jmrit/dispatcher/Dispatcher.shtml .  It might be possible to incorporate Java Help but I think a browser link would be nicer.  This does require that the user have a browser and an internet connection to display the help page.  Can the chm file be exported as html?  If so, it would be easy to merge the exported html into the standard JMRI help page template.
```
import java
import jmri
import webbrowser

webbrowser.open('https://www.jmri.org/help/en/package/jmri/jmrit/dispatcher/Dispatcher.shtml')
```
The URL would be modified to go directly to the help page.

After further research I discovered that you are not using the current PanelPro environment.  The second panel load is for direct script manipulation of the xml file.   This is risky because you will not be aware of changes that have occurred during the current JMRI session.

bill-fitch14 commented:
All I am doing is editing the xml of the panel to prodiuce another one, then asking the user to close the original panel and open th new one. I think I am using the normal panelpro techniques. I am sorry I had not made it clear that you have to close down the original panel. I have changed the instructions accordingly. I did not think it safe to edit the original panel directly.

Thanks I am now testing the help. I have researched and am using the JMRI help package instead of the webbrowser approach, but I did not know if it was allowed to put my help in the package area.

KenC57 commented:
Bill F,

It would seem better that you effectively copy, by creating a new panel file
and that is what gets edited. Then you have the user restart but use the new
panel file instead of the old one. Unless your change is PURELY additive,
you must do the restart to get the right results.
-ken c

dsand47 commented:
> Bill F, It would seem better that you effectively copy, by creating a new panel file and that is what gets edited. Then you have the user restart but use the new panel file instead of the old one. Unless your change is PURELY additive, you must do the restart to get the right results. -ken c

Ken,

The process loads the selected xml file independent of JMRI, insert the modifications, and writes a new xml file with a suffix.  

My concern is that  I start PanelPro, load A.xml and make a bunch of changes.  Then I run the script which loads A.xml and creates A.suffix.xml.  I then store A.xml and exit PanelPro.  When I load A.suffix.xml my changes are gone.

The potential for parallel changes is a risk.

dsand47 commented:
> Thanks I am now testing the help. I have researched and am using the JMRI help package instead of the webbrowser approach, but I did not know if it was allowed to put my help in the package area.

Since the process is dependent on Dispatcher and is used to add new capabilities, I think it is a good location.

bill-fitch14 commented:
Would a bit of code to save the current panel first before doing Stage1 solve the problem. With prompts of course so user knows what is happening.

bill-fitch14 commented:
At present (I would have to check) I think the changes to A.xml would be stored in A_icons.xml, but not in A.xml. Not a very good situation if the user did not like the new file.

bill-fitch14 commented:
I should really check before posting.
I loaded a.xml. Added  dummy block. Ran stage1 Saved A.xml
A.xml now does not have the edits from Stage1, but does have dummy block
A_icons.xml has the edits from Stage1 but not the dummy block.
I other words I do not change the xml of A, but load it into a DOM. modify the dom and save the DOM to A_icons.xml

I think aprompt to save A.xml might not be a bad idea, even though the user can save A.xml and redo Stage 1.

bill-fitch14 commented:
I do not change the loaded panel in memory in stage1. In stage2 I modify the panel in memory and save to a_icons.xml. You should not really modify a_icons.xml so I don't think you really need to save a_icons before modifying, but it won't hurt I suppose.

bill-fitch14 commented:
Correction In stage2 I modify the panel in memory and save to a_run.xml.

KenC57 commented:
Bill,

I am a big fan of multiple small steps with lots of savepoints so you can
track which step didn't go right.

dsand47 commented:
While there is no ideal solution to DOM vs. direct updates, I think direct updates with sufficient warnings will work best.  It keeps all of the changes in the JMRI context.  

What is the purpose of the layout block memory labels?  If they are only for block tracking, the block content labels are more efficient.

bill-fitch14 commented:
The layout block memory labels are for block tracking. I can change them to block content labels. Why are block content labels better. What are the memory labels for.

bill-fitch14 commented:
Direct Updates might be better, but would have been a lot more difficult to achieve and would have meant that a lot more modules would have had to be made public, something I was not sure would have been permisable. Perhaps if the whole approach is suitable it can be modified at a later stage and written in Java, by someone more able than myself. That would mean that the modules could remain private or protected.

The main disadvantage of xml modification as I see it is that it will need modifiation if the format of the xml changes. Hopefully the xml is fairly standardised at present.

bill-fitch14 commented:
Ken
The modifications that occur on the system either go right, or don't go at all. If they don't run the system crashes, then they can be debugged by changing the loglevel and putting in more comments, and finding the fault. I don't really think there would be any advantage in splitting the xml modifications into seperate stages, nor splitting the transit and traininfo generation as the latter are linked.

bill-fitch14 commented:
Sorry about the backslashes, I'll do them tomorrow.
I am struggling with the help. I have set up the help code but can't dispaly the help because i have a myriad of html files instead of one shtml file. I can't find a free converter. How does JMRI produce the shtml files? Probably they have bought something?

dsand47 commented:
> I am struggling with the help. I have set up the help code but can't dispaly the help because i have a myriad of html files instead of one shtml file. I can't find a free converter. How does JMRI produce the shtml files? Probably they have bought something?

The basic process is to make a copy of help/en/Template.shtml at the desired location and copy the content of the source html document into the shtml version with your favorite editor.  If the CHM html is like other MS html that I have seen, additional editing will be required to clean it up.

If you want, attach a zip file of the exported files and I can reformat them.

bill-fitch14 commented:
I have tried copying all the html help files into the package area and get a help mapping error when I run. One is DispatcherSystem.html. I tried putting DiapatcherSystem.html into jimrit.dispatcher and got the same error 

jmri.util.HelpUtil.helpMenu(bar, 'package.jmri.jmrit.dispatcherSystem.DispatcherSystem' , False)

javax.help.BadIDException: Not valid ID: package.jmri.jmrit.dispatcherSystem.DispatcherSystem
	at javax.help.Map$ID.create(Map.java:126)
	at javax.help.CSH.displayHelp(CSH.java:743)
	at javax.help.CSH.access$200(CSH.java:80)
	at javax.help.CSH$DisplayHelpFromSource.actionPerformed(CSH.java:1668)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)
Is there anything you recommend

bill-fitch14 commented:
Thanks. I have put a zipfile of the help into the Dispatcher System Files area. I would like to create the shtml myself, but I don't quite understand your explanation or know what tidying I would need to do. I would prefer not to bother you when the help needs updating. If there is a Linux or preferably another windows help package which would work better (not provide a advertisement watermark) let me know. Lets just see whether this works first though.

dsand47 commented:
> Thanks. I have put a zipfile of the help into the Dispatcher System Files area. I would like to create the shtml myself, but I don't quite understand your explanation or know what tidying I would need to do. I would prefer not to bother you when the help needs updating. If there is a Linux or preferably another windows help package which would work better (not provide a advertisement watermark) let me know. Lets just see whether this works first though.

See https://www.jmri.org/help/en/html/doc/Technical/Help.shtml for details on Help.  The **Creating the control files with JHelpDev** section provides the details for updating the Java Help index and table of contents.

My idea was to avoid the hassle of updating the Java Help control files.

The help pages are based on HTML 4 and CSS 1.  The  TemplatePlain.shtml provides the headers, footers and formatting for JMRI help pages.  The author inserts the desired content in `<div id="mainContent">`.  I use my program editor to manually create the html content.  

The attached text file contains the navigation tree and the Welcome content.  This will show you the basic structure of a JMRI help page.  You will notice that the actual format is quite simple.

Most people tend to put everything into one page with a navigation tree.  It is certainly possible to create lots of pages.  I would just keep adding `<h2>` and corresponding anchor, `<a name="???" id="???"></a>`, entries.

[DispatcherSystem.shtml.txt](https://github.com/JMRI/JMRI/files/5045432/DispatcherSystem.shtml.txt)

bill-fitch14 commented:
Changed help to shtml file but still got a help mapping error. I desperation I renamed dispatcher help to dispatcher-OLD.shtml, and put my help in place. My helpnow displays.
I have three help attempts:
jmri.util.HelpUtil.helpMenu(bar, 'package.jmri.jmrit.dispatcher.Dispatcher' , True)  #works but no index or seperate pages
jmri.util.HelpUtil.helpMenu(bar, 'package.jmri.jmrit.dispatcherSystem.DispatcherSystem' , False)    # does not work
jmri.util.HelpUtil.helpMenu(bar, 'package.jmri.jmrit.dispatcher.Dispatcher-OLD' , False)    # does not work
1) can you tell me what i shoud do to correct DispatcherSystem.shtml
2) is there anything that can be done with the mapping error, or does tat just not work on my system (Netbeans)

dsand47 commented:
> Changed help to shtml file but still got a help mapping error. I desperation I renamed dispatcher help to dispatcher-OLD.shtml, and put my help in place. My helpnow displays.
> I have three help attempts:
> jmri.util.HelpUtil.helpMenu(bar, 'package.jmri.jmrit.dispatcher.Dispatcher' , True) #works but no index or seperate pages
> jmri.util.HelpUtil.helpMenu(bar, 'package.jmri.jmrit.dispatcherSystem.DispatcherSystem' , False) # does not work
> jmri.util.HelpUtil.helpMenu(bar, 'package.jmri.jmrit.dispatcher.Dispatcher-OLD' , False) # does not work
> 
> 1. can you tell me what i shoud do to correct DispatcherSystem.shtml
> 2. is there anything that can be done with the mapping error, or does tat just not work on my system (Netbeans)

Since you appear to be set on using Java Help, you need to update the JHelpDev control files as described in https://www.jmri.org/help/en/html/doc/Technical/Help.shtml.  With this approach you can implement a single page or multiple pages.  Most people choose a single page with a navigation tree.  It is a lot easier.  This should fix the mapping errors.

Since the pages will also be accessible from jmri.org, they still need to converted using the JMRI template.  The shtml includes are ignored by Java Help so the pages work for either environment.  All of the extra coding needs to be removed.  The anchors need both the `name` and `id` attributes.  Java Help uses the name, browsers use the id.

The convention is that images are stored in an `image` directory, not `lib`.   

If you plan to have a standalone directory for the dispatcher system, it should be moved from the `package` tree to the `html` tree, maybe html/apps/DispatcherSystem.  Do this before doing the JHelpDev process.

bill-fitch14 commented:
This looks all very complicated. I tried the webbrowser, but did not know what url to put in. I put in the url provided by HelpNdoc and that worked,but then I was stuck.

bobjacobsen commented:
I think the solution to help mapping errors is just (macOS commands, but the Windows versions should be quite similar):

```
cd help/en
ant
```

The `ant` command will run the various steps needed to remake the various JavaHelp control files.

bill-fitch14 commented:
I have tried the webbrowser again, I have found that it has a peculiar effect. Even though I have put dispatcherSystem help in the dispatcher area, the dispatcher help is displayed.
    webbrowser.open('.../jmri/jmrit/dispatcher/Dispatcher.shtml') dispatcher help is displayed
    webbrowser.open('.../jmrit/dispatcher/Dispatcher-OLD.shtml') get The requested URL was not found on this server.   
    webbrowser.open('.../jmri/jmrit/dispatcherSystem/DispatcherSystem.shtml')  get The requested URL was not found on this server.

What do you think is the quickest easiest way to get the help going. I would preferably be able to edit the help by typing into a WYSIWYG editor and not doing too many edits, but not necessarily - we need JMRI compatible help.

dsand47 commented:
> I have tried the webbrowser again, I have found that it has a peculiar effect. Even though I have put dispatcherSystem help in the dispatcher area, the dispatcher help is displayed.
> webbrowser.open('.../jmri/jmrit/dispatcher/Dispatcher.shtml') dispatcher help is displayed
> webbrowser.open('.../jmrit/dispatcher/Dispatcher-OLD.shtml') get The requested URL was not found on this server.
> webbrowser.open('.../jmri/jmrit/dispatcherSystem/DispatcherSystem.shtml') get The requested URL was not found on this server.
> 
> What do you think is the quickest easiest way to get the help going. I would preferably be able to edit the help by typing into a WYSIWYG editor and not doing too many edits, but not necessarily - we need JMRI compatible help.

The webbrowser is using the real ULR to jmri.org which does not have your changes.

The problem with WYSIWYG editors is that they put all of their own CSS in which might conflict with JMRI CSS resulting in badly formed pages.  

The offer to assist is still available.  I can also do the Java Help changes.

bill-fitch14 commented:
Thanks very much Dave, offer accepted! I just hope that I don't have too many edits!

bill-fitch14 commented:
I will put your dispatcher back where it should go.I am just running ant, so I will run it again afterwards. I will include some edits that I was doing based on your suggestions, and recommit. Hopefully you won't have too much work. Thanks again.

dsand47 commented:
> I will put your dispatcher back where it should go.I am just running ant, so I will run it again afterwards. I will include some edits that I was doing based on your suggestions, and recommit. Hopefully you won't have too much work. Thanks again.

It would also be helpful if you moved dispatcherSystem and its contents from package/jmri/jmrit/ to html/apps/ and rename `lib` to `images`.  This will reduce the number of changes on my end.

bill-fitch14 commented:
the ant seems to have worked with the Java Help, but not with the webbrowser. I have moved the files. The links don't seem to have worked, nor are they on the lhs.
If there is any more I can do let me know, otherwise thanks.

dsand47 commented:
> the ant seems to have worked with the Java Help, but not with the webbrowser. I have moved the files. The links don't seem to have worked, nor are they on the lhs.
> If there is any more I can do let me know, otherwise thanks.

My changes have been uploaded to my repository: dsand47/JMRI.  The branch is ds_dispatcher_system.

According to @bobjacobsen you should be able to pull it into your local repository.  I have not tried such a thing before so I don't know the steps.

The branch includes the html cleanup, conversion to JMRI conventions, creating the Java Help connections, and modifications to the python program to use the standard JMRI Help menu which will link to the new help page.

I did not make any editorial changes.  It appears that the export from CHM file was incomplete.  There are sentences that seem to just stop.  

You still need to delete package.jmri.jrmit.DispatcherSystem and its contents.

bobjacobsen commented:
The git commands are:
```
git fetch https://github.com/dsand47/JMRI.git
git merge sand/ds_dispatcher_system
```

bill-fitch14 commented:
problems
bill (DispatcherSystem) JMRI-1 $ git fetch https://github.com/dsand47/JMRI.git
From https://github.com/dsand47/JMRI
 * branch                  HEAD       -> FETCH_HEAD
bill (DispatcherSystem) JMRI-1 $ git merge sand/ds_dispatcher_system
merge: sand/ds_dispatcher_system - not something we can merge

bill-fitch14 commented:
I went into git gui, loaded my repository, checked out the branch DispatcherSystem, went to gitbash from git gui, and ran the commands.

bobjacobsen commented:
Sorry, typo, I lost the “d” in dsand:

git merge dsand/ds_dispatcher_system

dsand47 commented:
Would the 47 also be needed?

----- Original message -----
From: Bob Jacobsen <notifications@github.com>
To: JMRI/JMRI <JMRI@noreply.github.com>
Cc: Dave Sand <gh@dfm.sandsys.org>, Manual <manual@noreply.github.com>
Subject: Re: [JMRI/JMRI] Dispatcher system (#8897)
Date: Monday, August 10, 2020 11:08 AM



Sorry, typo, I lost the “d” in dsand:

git merge dsand/ds_dispatcher_system




—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/JMRI/JMRI/pull/8897#issuecomment-671446174>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AED27YFYGB5F52OQQLUHBUDSAALPPANCNFSM4PVPPXPQ>.

bill-fitch14 commented:
bill (DispatcherSystem) JMRI-1 $ merge: dsand/ds_dispatcher_system - not something we can merge
bash: merge:: command not found
bill (DispatcherSystem) JMRI-1 $ git merge dsand47/ds_dispatcher_system
merge: dsand47/ds_dispatcher_system - not something we can merge

dsand47 commented:
> problems
> bill (DispatcherSystem) JMRI-1 $ git fetch https://github.com/dsand47/JMRI.git
> From https://github.com/dsand47/JMRI
> 
> * branch                  HEAD       -> FETCH_HEAD
>   bill (DispatcherSystem) JMRI-1 $ git merge sand/ds_dispatcher_system
>   merge: sand/ds_dispatcher_system - not something we can merge

Disclaimer:  I only know a tiny bit about git.

Try the following:
- git fetch https://github.com/dsand47/JMRI.git ds_dispatcher_system:ds_dispatcher_system
- git status -- See what is what.  Verify that the branch has been downloaded.
- git checkout `<yourbranch>`
- get merge ds_dispatcher_system

bobjacobsen commented:
> 
> bill (DispatcherSystem) JMRI-1 $ merge: dsand/ds_dispatcher_system - not something we can merge
> bash: merge:: command not found
> bill (DispatcherSystem) JMRI-1 $ git merge dsand47/ds_dispatcher_system
> merge: dsand47/ds_dispatcher_system - not something we can merge
> 

First line:  Would have been right, but the “git” is missing, no “:"

Second line:  Should not have 47, not sure where that came from.

Right syntax:

git merge dsand/ds_dispatcher_system

bill-fitch14 commented:
bill (DispatcherSystem) JMRI-1 $ git status
On branch DispatcherSystem
Your branch is up to date with 'origin/DispatcherSystem'.

nothing to commit, working tree clean
bill (DispatcherSystem) JMRI-1 $ git merge dsand/ds_dispatcher_system
merge: dsand/ds_dispatcher_system - not something we can merge
bill (DispatcherSystem) JMRI-1 $

bill-fitch14 commented:
bill (DispatcherSystem) JMRI-1 $ git fetch https://github.com/dsand47/JMRI.git
From https://github.com/dsand47/JMRI
 * branch                  HEAD       -> FETCH_HEAD
bill (DispatcherSystem) JMRI-1 $ git status
On branch DispatcherSystem
Your branch is up to date with 'origin/DispatcherSystem'.

nothing to commit, working tree clean
bill (DispatcherSystem) JMRI-1 $ git merge ds_dispatcher_system
merge: ds_dispatcher_system - not something we can merge
bill (DispatcherSystem) JMRI-1 $ git merge dsand/ds_dispatcher_system
merge: dsand/ds_dispatcher_system - not something we can merge
bill (DispatcherSystem) JMRI-1 $

dsand47 commented:
> bill (DispatcherSystem) JMRI-1 $ git fetch https://github.com/dsand47/JMRI.git
> From https://github.com/dsand47/JMRI
> 
> * branch                  HEAD       -> FETCH_HEAD
>   bill (DispatcherSystem) JMRI-1 $ git status
>   On branch DispatcherSystem
>   Your branch is up to date with 'origin/DispatcherSystem'.
> 
> nothing to commit, working tree clean
> bill (DispatcherSystem) JMRI-1 $ git merge ds_dispatcher_system
> merge: ds_dispatcher_system - not something we can merge
> bill (DispatcherSystem) JMRI-1 $ git merge dsand/ds_dispatcher_system
> merge: dsand/ds_dispatcher_system - not something we can merge
> bill (DispatcherSystem) JMRI-1 $

Try doing a `git branch`.  That will show what branches exist.

bobjacobsen commented:
My mistake.  The 47 does have to be there, Dave Sand’s repository is dsand47

git merge dsand47/ds_dispatcher_system



> bill (DispatcherSystem) JMRI-1 $ git fetch https://github.com/dsand47/JMRI.git
> From https://github.com/dsand47/JMRI
> 
> 	• branch HEAD -> FETCH_HEAD
> bill (DispatcherSystem) JMRI-1 $ git status
> On branch DispatcherSystem
> Your branch is up to date with 'origin/DispatcherSystem'.
> nothing to commit, working tree clean
> bill (DispatcherSystem) JMRI-1 $ git merge ds_dispatcher_system
> merge: ds_dispatcher_system - not something we can merge
> bill (DispatcherSystem) JMRI-1 $ git merge dsand/ds_dispatcher_system
> merge: dsand/ds_dispatcher_system - not something we can merge
> bill (DispatcherSystem) JMRI-1 $
>

bill-fitch14 commented:
bill (DispatcherSystem) JMRI-1 $ git merge ds_dispatcher_system
merge: ds_dispatcher_system - not something we can merge
bill (DispatcherSystem) JMRI-1 $ git merge dsand/ds_dispatcher_system
merge: dsand/ds_dispatcher_system - not something we can merge
bill (DispatcherSystem) JMRI-1 $ git merge dsand47/ds_dispatcher_system
merge: dsand47/ds_dispatcher_system - not something we can merge
bill (DispatcherSystem) JMRI-1 $ git branch
* DispatcherSystem
  master

What I might not have made clear is that I forked JMRI to JMRI-1 and created a branch DispatcherSystem

bill-fitch14 commented:
Maybe I should try this on the DispatcherSystem branch off JMRI instead of JMRI-1

bill-fitch14 commented:
Would that make a difference?

bobjacobsen commented:
We seem to have gotten to a strange git place.  Hard to debug remote.

Dave, may I make a PR to merge your branch to Bills?  Then he just has to click to merge.

dsand47 commented:
> We seem to have gotten to a strange git place. Hard to debug remote. Dave, may I make a PR to merge your branch to Bills? Then he just has to click to merge.

Yes.  Do whatever needs to be done to get Bill working again.

bobjacobsen commented:
When I do the merge manually, I get:
```
% git merge dsand-47-ds_dispatcher_system
Updating 6e352ca08c..e3322f3dca
Fast-forward
 help/en/JavaHelpSearch/DOCS                               |  Bin 759457 -> 781854 bytes
 help/en/JavaHelpSearch/DOCS.TAB                           |  Bin 17063 -> 17208 bytes
 help/en/JavaHelpSearch/OFFSETS                            |  Bin 10102 -> 10274 bytes
 help/en/JavaHelpSearch/POSITIONS                          |  Bin 14277330 -> 14795852 bytes
 help/en/JavaHelpSearch/SCHEMA                             |    2 +-
 help/en/JavaHelpSearch/TMAP                               |  Bin 208896 -> 210944 bytes
 help/en/JmriHelp_enIndex.xml                              |    2 +
 help/en/JmriHelp_enTOC.xml                                |    2 +
 help/en/Map.jhm                                           |  574 ++++++-------
 help/en/html/apps/DispatcherSystem/DispatcherSystem.shtml |  456 ++++------
 help/en/webindex.shtml                                    | 4349 ++++++++++++++++++++++++++++++++++++++++++++++++++--------------------------------------------
 help/en/webtoc.shtml                                      | 4807 +++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------------
 jython/DispatcherSystem/DispatcherSystem.py               |  100 +--
 13 files changed, 5414 insertions(+), 4878 deletions(-)
```

Does that look right?

bobjacobsen commented:
@bill-fitch14  Please go to "https://github.com/bill-fitch14/JMRI-1/pull/1" and click the Merge button at the bottom.  It might ask you to confirm, in which case please do.  

That will merge @dsand47's changes _on Github_.  That in turn will update them in this PR.  If that makes it ready to merge, we're all done.

If you want to made further changes, you'll need to pull this back onto your computer.  Let me know and I'll walk you through it.

(The command line problems appear to be due to a confusion somewhere in the local history between https://github.com/bill-fitch14/JMRI-1.git and https://github.com/bill-fitch14/JMRI.git - once we get this merged, that won't matter any more)

dsand47 commented:
> When I do the merge manually, I get:
> 
> ```
> % git merge dsand-47-ds_dispatcher_system
> Updating 6e352ca08c..e3322f3dca
> Fast-forward
>  help/en/JavaHelpSearch/DOCS                               |  Bin 759457 -> 781854 bytes
>  help/en/JavaHelpSearch/DOCS.TAB                           |  Bin 17063 -> 17208 bytes
>  help/en/JavaHelpSearch/OFFSETS                            |  Bin 10102 -> 10274 bytes
>  help/en/JavaHelpSearch/POSITIONS                          |  Bin 14277330 -> 14795852 bytes
>  help/en/JavaHelpSearch/SCHEMA                             |    2 +-
>  help/en/JavaHelpSearch/TMAP                               |  Bin 208896 -> 210944 bytes
>  help/en/JmriHelp_enIndex.xml                              |    2 +
>  help/en/JmriHelp_enTOC.xml                                |    2 +
>  help/en/Map.jhm                                           |  574 ++++++-------
>  help/en/html/apps/DispatcherSystem/DispatcherSystem.shtml |  456 ++++------
>  help/en/webindex.shtml                                    | 4349 ++++++++++++++++++++++++++++++++++++++++++++++++++--------------------------------------------
>  help/en/webtoc.shtml                                      | 4807 +++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------------
>  jython/DispatcherSystem/DispatcherSystem.py               |  100 +--
>  13 files changed, 5414 insertions(+), 4878 deletions(-)
> ```
> 
> Does that look right?

Yes.  An updated help page, an update using JavaHelpDev and changing the python program to use the standard JMRI Help menu.

bill-fitch14 commented:
I have merged, and Dave's edits are in DispatcherSystem.shtml. Thanks a lot both of you. I will check the help to correct the faults Dave has picked up.

bill-fitch14 commented:
I have I hpe completed the help edits. Hope everything is still working after the latest commit.Do I have to do anything about the conflicting file help/en/JmriHelp_enTOC.xml

bobjacobsen commented:
I fixed that conflict.  

You should pull back the branch from GitHub to pick that up (like after you merged the PR earlier)

bobjacobsen commented:
@dsand47 I think it would make sense to merge this.  Even if there’s more to do, that would give us a common base. OK?

dsand47 commented:
> @dsand47 I think it would make sense to merge this. Even if there’s more to do, that would give us a common base. OK?

I agree.  There are a couple of minor cleanup items but they can wait.  It would be good to get the JavaHelp changes merged.

bobjacobsen commented:
OK.  @bill-fitch14 thanks for this!

bill-fitch14 commented:
Thankyou for your help. I have deleted the branch. Presumably any new edits should be done with a new branch.

bill-fitch14 commented:
I can restore it if necessary

bobjacobsen commented:
That's fine.  When you refresh your repository contents ("git pull"), pull pick up your changes and everybody else's in the current master branch, and can build on that with a new branch.

And thanks again for this. It looks cool.
